badge_server/main.py

```diff
@@ -137,24 +138,53 @@ def get_highest_status(cls, statuses: Iterable[enum.Enum]):
 }
 
 
-def _get_supported_versions(package_name: str) -> List[int]:
-    """Gets the given package's supported python versions
+def _get_missing_details(package_names: List[str],
+                         results: Iterable[CompatibilityResult]) -> str:
+    """Gets the details for any missing data (if there is any)
 
     Args:
-        package_name: The name of the package to look up, e.g. "tensorflow".
+        package_names: A list of length 1 or 2 of the package name(s) to look
+            up, e.g. ["tensorflow"], ["tensorflow", "opencensus"].
+        results: A list of length 1 or 2 of the `CompatibilityResults` for the
+            given package(s).
 
     Returns:
-        A list of supported python versions.
+        None if there is no missing data; a description of the version(s)
+        missing otherwise, e.g. "Missing data for python version(s): 2".
     """
-    supported = []
-    if not compat_utils._is_package_in_whitelist([package_name]):
-        return supported
+    expected_lengths = (1, 2)
+    bad_list_length = 'package_names: Expected length of 1 or 2, got %s'
+    not_whitelisted = 'One of the packages in %s is not whitelisted'
+
+    if len(package_names) not in expected_lengths:
+        # This should not happen
+        raise ValueError(bad_list_length % len(package_names))
+
+    if not compat_utils._is_package_in_whitelist(package_names):
+        # This should not happen
+        raise ValueError(not_whitelisted % str(package_names))
+
+    versions = (2, 3)
```

Comment:
I chose to use a tuple, since the data inside would be static. Why use a list?

Note: I changed the variable name to `all_versions`

Reply:
In Python, tuples tend to be used for heterogenous collections and list for homogeneous collections:
https://mail.python.org/pipermail/python-dev/2014-April/134140.html
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#3199-tuples-vs-lists

Reply:
I honestly don't mind using a list instead, but this my first time hearing about homogeneity/heterogeneity being a distinction and I'd like determine if this should actually be a best practice or not because that would affect me going forward. Personally, the only distinction I've heard of is that tuples are immutable while lists are mutable. This discussion seems to back that up: https://bugs.python.org/issue14840 unless I'm misinterpreting things/am missing some other context.

Specifically these ones:
"Im/mutability is part of the definition and operation of the language. Homo/heter/geneity is not (that I can think of at the moment)."

"there are parts of Python's own syntax that
require a tuple that happens to be a homogenous sequence, such as the
:keyword:`except` clause of a ``try ... except`` statement, :func:`issubclass`,
and :func:`isinstance`"

Reply:
OK, I think that I retract my comment - both patterns seem equally common in Google3. I started a discussion in the team chat ;-)

Reply:
I don't think I'm in the team chat.

Reply:
I just added you.